Sources/PenguinStructures/SourceInitializableCollection.swift

```diff
@@ -0,0 +1,26 @@
+//******************************************************************************
+// Copyright 2020 Penguin Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+/// A collection that can be initialized to contain exactly the elements of a source collection.
+public protocol SourceInitializableCollection: Collection {
```

Comment:
How is this not just `RangeReplaceableCollection`?  I realize that protocol has more operations, but for this to be a legit separate protocol there need to be models for which we *cannot* reasonably implement the requirements of `RangeReplaceableCollection`.  The fact that we haven't done it for `ArrayBuffer` doesn't mean anything.

I see; you're using it for `FixedSizedArray`.  Okay, but then there should be some precondition that you can talk about that describes how we know the initialization is legal, since it will have to trap in the other cases.  

The other thing any protocol needs in order to be legit is a family of generic algorithms that depend on it.  I don't see any algorithms depending on `SourceInitializableCollection`.

Reply:
> I see; you're using it for FixedSizedArray. Okay, but then there should be some precondition that you can talk about that describes how we know the initialization is legal, since it will have to trap in the other cases.

Or I could make it failable? But then I would have to make FixedSizeArray's `init<Source>(_ source: Source)` failable. Would that be ok?

> The other thing any protocol needs in order to be legit is a family of generic algorithms that depend on it. I don't see any algorithms depending on SourceInitializableCollection.

The thing I am trying to do is a "map" with `SourceInitializableCollection` return type.

Concretely, [here](https://github.com/borglab/SwiftFusion/blob/8a2ec9fd4b2fd07ff0e2c05b1b20016b6b8aae4e/Sources/SwiftFusion/Inference/JacobianFactor.swift#L50) I need to construct an array of pullbacks of the standard basis vectors, and I usually know statically how many basis vectors there are so it's most efficient to store them in a fixed size array.

Does this sound like the beginning of a family of generic algorithms that makes it legit? Or is there some other way to achieve what I'm doing?

Reply:
`Rows` is a `FixedSizeArray`.  I think you can rebuild `FixedSizeVector` on `FixedSizeArray` and then provide (shadow) algorithms like `map` that produce only arrays of the same statically-known size.

Reply:
It's a `FixedSizeArray` in `master` right now, but my PR https://github.com/borglab/SwiftFusion/pull/148 turns it into a `SourceInitializableCollection` so that it can also be `Array`.